File: starter_kits/Ruby/hlt/game.rb
require 'logger'
require 'json'
require 'constants'
require 'map'
require 'player'

# The game object holds all metadata pertinent to the game and all its contents
class Game
  attr_reader :map, :me, :turn_number, :my_id

  # Initiates a game object collecting all start-state instances for the contained items for pre-game.
  # Also sets up basic logging.
  def initialize(name)
    @name = name
    # implicit IO flush
    $stdout.sync = true

    # Grab constants JSON
    Constants.load_constants(JSON.parse(read_from_input))

    @turn_number = 0
    @num_players, @my_id = read_ints_from_input

    @players = {}
    @num_players.times do
      player_id, player = Player.generate(read_ints_from_input)
      @players[player_id] = player
    end

    @me = @players[@my_id]
    @map = Map.generate(self)
  end

  # Indicate that your bot is ready to play.
  def ready
    end_turn([@name])
  end

  # Updates the game object's state.
  # :returns: nothing.
  def update_frame
    @turn_number = read_ints_from_input.first
    LOGGER.info("=============== TURN #{@turn_number} ================")

    @num_players.times do |_|
      player, num_ships, num_dropoffs, halite = read_ints_from_input
      @players[player].update(self, num_ships, num_dropoffs, halite)
    end

    @map.update(self)


    # Mark cells with ships as unsafe for navigation
    for player in @players.values
      for ship in player.ships
        @map[ship.position].mark_unsafe(ship)
      end

      @map[player.shipyard.position].structure = player.shipyard
      for dropoff in player.dropoffs
        @map[dropoff.position].structure = dropoff
      end
    end
  end

  def end_turn(commands)
    write_to_output(commands.join(" "))
  end

  def read_from_input
    $stdin.gets.strip
  end

  def read_ints_from_input
    read_from_input.split(' ').map { |v| Integer(v) }
  end

  private

  def write_to_output(data)
    data = "#{data.strip}"
    LOGGER.info("Sending: #{data.inspect}")
    $stdout.puts(data)
  end
end

File: starter_kits/TypeScript/hlt/Game.ts
import { Dropoff } from "./Dropoff";
import { GameMap } from "./GameMap";
import { Logging } from "./Logging";
import { Player } from "./Player";
import { Position } from "./Position";
import { ServerCommunication } from "./ServerCommunicaion";
import { Ship } from "./Ship";
import { Shipyard } from "./Shipyard";

export class Game {
    public turnNumber = 0;
    public server: ServerCommunication = new ServerCommunication();

    public myId = 0;
    public players = new Map<number, Player>();
    public me?: Player;
    public gameMap?: GameMap;

    /**
     * Initialize a game object collecting all the start-state
     * instances for pre-game. Also sets up a log file in
     * "bot-<bot_id>.log".
     * @returns The initialized gameMap and me so we don't have to check if undefined.
     */
    public async initialize(): Promise<[GameMap, Player]> {
        const serverData = await this.server.getInitialData();
        this.myId = serverData.myId;

        Logging.setup(`bot-${this.myId}.log`);

        serverData.players.forEach((playerData) => {
            const player = new Player(playerData.id,
                new Shipyard(playerData.id, -1, new Position(playerData.x, playerData.y)));
            this.players.set(player.id, player);
        });
        this.me = this.players.get(this.myId);

        this.gameMap = new GameMap(serverData.cells);

        return [this.gameMap as GameMap, this.me as Player]; // We cast here because we have just initialized
    }

    /**
     * Updates the game object's state.
     */
    public async updateFrame() {
        const data = await this.server.getUpdateData(this.players.size);
        this.turnNumber = data.turn;
        Logging.info(`================ TURN ${this.turnNumber.toString().padStart(3, "0")} ================`);
        data.players.forEach((playerData) => {
            const player = this.players.get(playerData.id) as Player;
            player.haliteAmount = playerData.halite;

            // Process ships
            const newShipsData = playerData.ships
                .filter((shipData) => !player.hasShip(shipData.id));
            newShipsData.forEach((newShipData) =>
                player.addShip(new Ship(player.id, newShipData.id,
                    new Position(newShipData.x, newShipData.y), newShipData.halite)));

            const lostShips = player.getShips()
                .filter((ship) => !playerData.ships.some((shipData) => shipData.id === ship.id));
            lostShips.forEach((ship) => player.loseShip(ship.id));

            player.getShips().forEach((ship) => {
                const updatedShipData = playerData.ships
                    .find((shipData) => ship.id === shipData.id);
                if (updatedShipData) {
                    [ship.haliteAmount, ship.position.x, ship.position.y] =
                        [updatedShipData.halite, updatedShipData.x, updatedShipData.y];
                }
            });

            // Process dropoffs
            const newDropoffsData = playerData.dropoffs
                .filter((dropoffData) => !player.dropoffs.has(dropoffData.id));
            newDropoffsData.forEach((newDropoffData: { id: number, x: number, y: number }) =>
                player.dropoffs.set(newDropoffData.id,
                    new Dropoff(player.id, newDropoffData.id, new Position(newDropoffData.x, newDropoffData.y))));

            const lostDropoffs = Array.from(player.dropoffs.values())
                .filter((dropoff) => !playerData.dropoffs.some((dropoffData) => dropoffData.id === dropoff.id));
            lostDropoffs.forEach((lostDropoff) => {
                player.dropoffs.delete(lostDropoff.id);
                player.lostDropoffs.set(lostDropoff.id, lostDropoff);
            });
        });

        const gameMap = this.gameMap as GameMap;
        // Mark all cells as safe
        gameMap.cells.forEach((row) => row.forEach((cell) => cell.markSafe()));
        // Update cells
        data.cells.forEach((cell) => gameMap.get(new Position(cell.x, cell.y)).haliteAmount = cell.halite);
        // Mark cells with ships as unsafe for navigation, mark sturctures
        for (const player of this.players.values()) {
            player.getShips()
                .forEach((ship) => gameMap.get(ship.position).markUnsafe(ship));
            player.getDropoffs()
                .forEach((dropoff) => gameMap.get(dropoff.position).structure = dropoff);
        }
    }

    /**
     * Indicate that your bot is ready to play by sending the bot name.
     */
    public async ready(botName: string) {
        await this.server.sendCommands([botName]);
    }

    /**
     * Send all commands to the game engine, effectively ending your
     * turn.
     */
    public async endTurn(commands: string[]) {
        await this.server.sendCommands(commands);
    }
}
